Update contributing.md with current processes (#2377)
AzureAD/microsoft-authentication-library-for-js: contributing.md

In case of doubt, open an [issue](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues).

Especially do so if you plan to work on a major change in functionality.  Nothing is more frustrating than seeing your hard work go to waste because your vision does not align with our goals for the SDK.

### Create a branch for your feature/patch

Okay, so you have decided on the proper branch.  Create a feature branch and start hacking:

```
$ git checkout -b my-feature-branch
```

## Installation and Bootstrapping

This repository is a monorepo for all packages maintained by the MSAL.js team. It contains public `npm` packages in the [`lib`](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib) folder as well as some additional libraries in the [`extensions`](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions) folder.

If you would like to build a specific package, you can visit the `README.md` for that specific folder to review the build process. The instructions here are specific to the monorepo. 

> After you have cloned the repository, run `npm install` in the root folder. This will install the depedencies for all of the libraries and samples in the repo, create linkages between packages that have depedencies on each other, and run the build process for each of the libraries.

Reviewer: ```suggestion
After you have cloned the repository, run `npm install` in the root folder. This will install the dependencies for all of the libraries and samples in the repo, create links between packages that depend on each other, and run the build process for each of the libraries.
```

Author: done.

> **NOTE**: This requires the installation of the `lerna` package to successfully run. You can find more information about `lerna` [here](https://www.npmjs.com/package/lerna).

> To reset each of the libraries, run `npm run clean` in the root folder, which will remove the `node_modules` folders and built files for each repo and sample.

Reviewer: ```suggestion
To reset each of the libraries, run `npm run clean` in the root folder, which will remove the `node_modules` folders and build files for each subrepo and sample.
```

Author: done.

If you run `npm install` in any of the libraries individually, you may find that your packages are no longer referencing the local files, but instead looking for the ones available publicly on `npm`. You can resolve this by running `lerna bootstrap`.

## Build and Test

### Build Library

> ```javascript

Reviewer: ```suggestion
```bash
```

Author: done.

// Change to the root of the msal repo
cd microsoft-authentication-library-for-js/
// Install npm dependencies and bootstrap packages
npm install
// Build library
npm run build
// To run build only for a single package
npm run build -- --scope <package-name>
```

### Test and Test Coverage

> Bug fixes and features should come with tests.  Add your tests in the

Reviewer: ```suggestion
Bug fixes and features should come with automated tests.  Add your tests in the
```

Author: done.

test directory. This varies by repository but often follows the same convention of /src/test.  Look at other tests to see how they should be
structured (license boilerplate, common includes, etc.).

> Make sure that all tests pass. You can run the tests as described below.

Reviewer: ```suggestion
Make sure that all previously passing and new tests pass before submitting your changes. You can run the tests as described below.
```

#### Unit tests without coverage:

> ```javascript

Reviewer: ```suggestion
```bash
```

Author: done.

### Create a branch for your feature/patch

Okay, so you have decided on the proper branch.  Create a feature branch and start hacking:

```
$ git checkout -b my-feature-branch
```

## Installation and Bootstrapping

This repository is a monorepo for all packages maintained by the MSAL.js team. It contains the following public npm packages:

- [msal](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-core)
- [@azure/msal](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-core) (same package as above)
- [@azure/msal-angular](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-angular)
- [@azure/msal-common](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-common)
- [@azure/msal-browser](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-browser)
- [@azure/msal-node](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node)

> - [@azure/msal-node-extensions](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/msal-node-extensions)

Reviewer 2: I might just link to the `lib` folder, so we don't have to keep this up to date.

Author: updated. Need to link to lib and extensions.

$ git config --global user.email "j.random.user@example.com"
```

Writing good commit logs is important.  A commit log should describe what changed and why.  Follow these guidelines when writing one:

> - 50 characters or less 

Reviewer 2: If this isn't enforced by a tool, I don't think its worth putting here.

Author: More of a general guideline, but agreed it's a bit arbitrary. Removing.

Checks that changefiles have been generated for relevant changed packages. Used by the CI tool on PRs, can also be used locally.

- `npm run beachball:bump`

> Bumps versions for packages with relevant changefiles. Run on `master` branch to check which packages are slated for release.

Reviewer 3: Maybe add a note that this should not be done when submitting a PR?

Author: Good point, will add it.

Reviewer: If we want the PR number in the change file message, we should suggest they should first create the PR, then run beachball:change, add the PR number in the message and then push to their feature branch again?

Author: I've added a comment for the PR number in the description for the `beachball:change` command.

Bumps versions and publishes all packages that have relevant changefiles. If the changefiles types are set to `None`, no version bump or publish is done. Run on push to the `master` branch.
Bumps versions for packages with relevant changefiles. Run on `master` branch to check which packages are slated for release.

- Notes

>    - Ensure you have the correct email set in your local git config.

Reviewer 3: This is good info to keep around. I would leave it in personally

Author: It's listed in one of the sections above.
